libs/contrib/Data/Elem/Extra.idr

```diff
@@ -0,0 +1,52 @@
+module Data.Elem.Extra
+
+import Data.List
+import Data.List.Elem
+
+%default total
+
+||| Proof that an element is still inside a list if we append to it.
+public export
+elemAppLeft : (xs, ys : List a) -> (prf : Elem x xs) -> Elem x (xs ++ ys)
+elemAppLeft (x :: xs) ys Here = Here
+elemAppLeft (x :: xs) ys (There prf2) = There $ elemAppLeft xs ys prf2
+
+
+||| Proof that an element is still inside a list if we prepend to it.
+public export
+elemAppRight :  (ys, xs : List a) -> (prf : Elem x xs) -> Elem x (ys ++ xs)
+elemAppRight [] xs prf = prf
+elemAppRight (y :: ys) xs prf = There $ elemAppRight ys xs prf
+
+||| Proof that membership on append implies membership in left or right sublist.
+public export
+elemAppLorR : (xs, ys : List a)
+           -> (prf : Elem k (xs ++ ys))
+           -> Either (Elem k xs) (Elem k ys)
+elemAppLorR [] [] prf = absurd prf
+elemAppLorR [] _ prf = Right prf
+elemAppLorR (x :: xs) [] prf =
+  let eq   : (xs ++ [] = xs)  = appendNilRightNeutral xs
+      prf' : Elem k (x :: xs) = replace {p = \g => Elem k (x :: g)} eq prf
+  in Left prf'
```

Comment:
By the way, this case could be written simpler:
```idris
elemAppLorR (x :: xs) [] prf = Left rewrite sym $ appendNilRightNeutral xs in prf
```

Reply:
I'm of two minds about this. On the one hand I think it's useful for people to see that rewrites are just an ordinary expression, on the other I also think there are not very many examples of `replace` floating about.
Should I be aiming for the minimal text possible when submitting proofs?

Reply:
> I think it's useful for people to see that rewrites are just an ordinary expression

My personal opinion is that standard libraries are not the best place for putting education code snippets. If you want clarify that `rewrite` is just a specially tuned `replace` expression, I think, you should add more clarification of it to the documentation rather than using it in the library just to show.

> Should I be aiming for the minimal text possible when submitting proofs?

Well, I don't know. I think that the proof that is understandable to most educated people is the best. *Personally* I think that short one with `rewrite` reads better than several `let`s -- but I'm sure there are people who think differently. Again, *personally* I think that `let`s add clarification only when expression is really big or where there are some similar expressions with completely different semantics -- in this case you'd want to attach a name to them. In all other cases, I think, `let`s do not add anything to readability and even disturb.